lib/dialect/include/rlc/dialect/ConstraintsAnalysis.hpp

```diff
@@ -0,0 +1,299 @@
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#pragma once
+
+#include <utility>
+#include <iostream>
+#include <limits>
+#include <optional>
+#include <string>
+
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Analysis/DataFlow/ConstantPropagationAnalysis.h"
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
+#include "mlir/Analysis/DataFlow/DenseAnalysis.h"
+#include "mlir/IR/BuiltinDialect.h"
+#include "mlir/Interfaces/ControlFlowInterfaces.h"
+#include "rlc/dialect/Operations.hpp"
+#include "rlc/dialect/Passes.hpp"
+#include "rlc/dialect/conversion/TypeConverter.h"
+#include "llvm/IR/ConstantRange.h"
+
+// Add the map key to the llvm scope
+// I have to admit that if this works the llms will take over the world someday
+// Ok it compiles
+// It seems to work. Wow
+namespace llvm{
+	template <>
+	struct llvm::DenseMapInfo<std::pair<mlir::Value,bool>>{
+		static std::pair<mlir::Value,bool> getEmptyKey(){
+			return std::make_pair(llvm::DenseMapInfo<mlir::Value>::getEmptyKey(),false);
+		}
+		static std::pair<mlir::Value, bool> getTombstoneKey() {
+			return std::make_pair(DenseMapInfo<mlir::Value>::getTombstoneKey(), false);
+		}
+		static unsigned getHashValue(const std::pair<mlir::Value, bool> &Val) {
+			return hash_combine(DenseMapInfo<mlir::Value>::getHashValue(Val.first), Val.second);
+		}
+		static bool isEqual(const std::pair<mlir::Value, bool> &lhs, const std::pair<mlir::Value, bool> &rhs) {
+			return DenseMapInfo<mlir::Value>::isEqual(lhs.first, rhs.first) && lhs.second == rhs.second;
+		}
+	};
+}
+
+namespace mlir::rlc
+{
+	
+	// Useful static methods
+
+	static bool findInterestingOperandAndUpdate(
+		mlir::Operation* op,
+		mlir::rlc::ConstraintsLattice* currentLattice,
+		mlir::rlc::ConstraintsAnalysis* analysis, 
+		const mlir::Value& result);
+
+	static bool updateArithmeticRange(
+		mlir::Operation* op,
+		mlir::rlc::ConstraintsLattice* currentLattice,
+		mlir::rlc::ConstraintsAnalysis* analysis, 
+		const mlir::Value& result,
+		const mlir::Value& update,
+		const mlir::Value& other,
+		bool branch);
+
+	static bool visitYield(
+		mlir::Operation* statement, 
+		mlir::rlc::ConstraintsLattice* currentLattice);
+
+	static bool visitUninitializedConstruct(
+		mlir::Operation* statement, 
+		mlir::rlc::ConstraintsLattice* currentLattice);
+
+	// Class prototypes
+	// LEAVE THEM OR IT WILL BREAK THE COMPILATION DUE TO A CIRCULAR DEPENDENCE WHICH FOR THE MOMENT CANNOT BE SOLVED (IT JUST WORKS cit.)
+	class ConstraintsLattice;
+	class ConstraintsAnalysis;
+
+	// I will leave this here
+
+	// AbstractDenseLattice
```

Comment:
drop comments there for you to understand things, only comments about explaining the code should be left.